File: mcp/src/velum_mcp/tools/tasks.py
"""MCP tools — Task management, execution, output, filtering."""

from __future__ import annotations

from typing import Any, Literal

from velum_mcp.client import get_client


async def list_tasks(
    project_id: int,
    limit: int = 20,
    status: Literal["waiting", "running", "success", "error", "stopped"] | None = None,
) -> list[dict[str, Any]]:
    """List recent tasks in a project, optionally filtered by status.

    Args:
        project_id: Project to query.
        limit: Maximum number of tasks to return (default 20).
        status: Filter by task status (optional).
    """
    client = get_client()
    tasks: list[dict[str, Any]] = await client.get(f"/projects/{project_id}/tasks") or []
    if status:
        tasks = [t for t in tasks if t.get("status") == status]
    return tasks[:limit]


async def get_task(project_id: int, task_id: int) -> dict[str, Any]:
    """Get full details of a specific task including status, timing, and metadata."""
    client = get_client()
    return await client.get(f"/projects/{project_id}/tasks/{task_id}")


async def run_task(
    project_id: int,
    template_id: int,
    message: str = "",
    git_branch: str | None = None,
    extra_vars: str | None = None,
    inventory_id: int | None = None,
) -> dict[str, Any]:
    """Run a task from a template. Returns the new task object with its ID.

    Args:
        project_id: Project ID.
        template_id: Template to run.
        message: Human-readable note for this run (shows in history).
        git_branch: Override branch/tag/commit from template defaults.
        extra_vars: JSON or YAML extra_vars override string.
        inventory_id: Override the template's default inventory.

    Returns:
        Task object — use task_id to poll status or stream logs.
    """
    client = get_client()
    body: dict[str, Any] = {"template_id": template_id}
    if message:
        body["message"] = message
    if git_branch:
        body["git_branch"] = git_branch
    if extra_vars:
        body["arguments"] = extra_vars
    if inventory_id is not None:
        body["inventory_id"] = inventory_id
    return await client.post(f"/project/{project_id}/tasks", body)


async def stop_task(project_id: int, task_id: int) -> str:
    """Send a stop signal to a running task.

    Args:
        project_id: Project ID.
        task_id: Task to stop.
    """
    client = get_client()
    await client.post(f"/projects/{project_id}/tasks/{task_id}/stop")
    return f"Stop signal sent to task {task_id}."


async def get_task_output(
    project_id: int,
    task_id: int,
    raw: bool = False,
    last_n_lines: int | None = None,
) -> str:
    """Retrieve the console output of a completed or running task.

    Args:
        project_id: Project ID.
        task_id: Task to read output from.
        raw: If True, return raw unformatted text; otherwise formatted records.
        last_n_lines: Return only the last N lines of output (optional).
    """
    client = get_client()
    path = f"/projects/{project_id}/tasks/{task_id}/raw_output" if raw else \
           f"/projects/{project_id}/tasks/{task_id}/output"
    result = await client.get(path)

    if isinstance(result, str):
        lines = result.splitlines()
    elif isinstance(result, list):
        lines = [r.get("output", "") for r in result]
    else:
        lines = [str(result)]

    if last_n_lines:
        lines = lines[-last_n_lines:]
    return "\n".join(lines)


async def filter_tasks(
    project_id: int,
    template_id: int | None = None,
    status: Literal["waiting", "running", "success", "error", "stopped"] | None = None,
    limit: int = 50,
) -> list[dict[str, Any]]:
    """Filter project tasks by template and/or status.

    Args:
        project_id: Project to query.
        template_id: Filter to tasks from a specific template (optional).
        status: Filter by execution status (optional).
        limit: Maximum results.
    """
    client = get_client()
    tasks: list[dict[str, Any]] = await client.get(f"/projects/{project_id}/tasks") or []
    if template_id is not None:
        tasks = [t for t in tasks if t.get("template_id") == template_id]
    if status:
        tasks = [t for t in tasks if t.get("status") == status]
    return tasks[:limit]


async def get_latest_failed_task(project_id: int) -> dict[str, Any] | str:
    """Return the most recent failed task in a project, or a message if none found."""
    client = get_client()
    tasks: list[dict[str, Any]] = await client.get(f"/projects/{project_id}/tasks") or []
    failed = [t for t in tasks if t.get("status") == "error"]
    if not failed:
        return "No failed tasks found in this project."
    # Sort by ID descending (higher ID = more recent)
    return max(failed, key=lambda t: t.get("id", 0))


async def get_waiting_tasks(project_id: int) -> list[dict[str, Any]]:
    """Return all tasks currently waiting in the queue for a project."""
    client = get_client()
    tasks: list[dict[str, Any]] = await client.get(f"/projects/{project_id}/tasks") or []
    return [t for t in tasks if t.get("status") == "waiting"]


async def bulk_stop_tasks(project_id: int, task_ids: list[int]) -> str:
    """Stop multiple tasks at once.

    Args:
        project_id: Project containing the tasks.
        task_ids: List of task IDs to stop.
    """
    client = get_client()
    results = []
    for tid in task_ids:
        try:
            await client.post(f"/projects/{project_id}/tasks/{tid}/stop")
            results.append(f"✓ {tid}")
        except Exception as e:  # noqa: BLE001
            results.append(f"✗ {tid}: {e}")
    return "Stop results:\n" + "\n".join(results)


async def confirm_task(project_id: int, task_id: int) -> str:
    """Confirm (approve) a task that is waiting for human approval (gated execution)."""
    client = get_client()
    await client.post(f"/projects/{project_id}/tasks/{task_id}/confirm")
    return f"Task {task_id} confirmed and will proceed."


async def reject_task(project_id: int, task_id: int) -> str:
    """Reject a task waiting for human approval — it will be stopped."""
    client = get_client()
    await client.post(f"/projects/{project_id}/tasks/{task_id}/reject")
    return f"Task {task_id} rejected and stopped."

File: mcp/src/tools/tasks.rs
use crate::client::VelumClient;
use crate::protocol::{prop_bool, prop_int, prop_int_opt, prop_str_opt, Tool, ToolResult};
use serde_json::{json, Value};
use anyhow::Result;

pub fn definitions() -> Vec<Tool> {
    vec![
        Tool {
            name: "list_tasks".into(),
            description: "List recent tasks in a project, optionally filtered by status.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "limit": prop_int_opt("Max tasks to return (default 20)"),
                    "status": { "type": ["string","null"], "enum": ["waiting","running","success","error","stopped",null], "description": "Filter by status" }
                },
                "required": ["project_id"]
            }),
        },
        Tool {
            name: "get_task".into(),
            description: "Get full details of a task including status, timing, and metadata.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "task_id": prop_int("Task ID")
                },
                "required": ["project_id","task_id"]
            }),
        },
        Tool {
            name: "run_task".into(),
            description: "Run a task from a template. Returns task object with ID for status tracking.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "template_id": prop_int("Template to run"),
                    "message": prop_str_opt("Human note for this run"),
                    "git_branch": prop_str_opt("Override branch/tag/commit"),
                    "arguments": prop_str_opt("Extra variables override (JSON string)"),
                    "inventory_id": prop_int_opt("Override inventory ID")
                },
                "required": ["project_id","template_id"]
            }),
        },
        Tool {
            name: "stop_task".into(),
            description: "Send a stop signal to a running task.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "task_id": prop_int("Task to stop")
                },
                "required": ["project_id","task_id"]
            }),
        },
        Tool {
            name: "get_task_output".into(),
            description: "Retrieve console output of a task. Optionally return only the last N lines.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "task_id": prop_int("Task ID"),
                    "raw": prop_bool("Return raw unformatted text (default: false)"),
                    "last_n_lines": prop_int_opt("Return only the last N lines")
                },
                "required": ["project_id","task_id"]
            }),
        },
        Tool {
            name: "filter_tasks".into(),
            description: "Filter project tasks by template and/or execution status.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "template_id": prop_int_opt("Filter to a specific template"),
                    "status": { "type": ["string","null"], "enum": ["waiting","running","success","error","stopped",null] },
                    "limit": prop_int_opt("Max results (default 50)")
                },
                "required": ["project_id"]
            }),
        },
        Tool {
            name: "get_latest_failed_task".into(),
            description: "Return the most recent failed task in a project for quick error inspection.".into(),
            input_schema: json!({
                "type": "object",
                "properties": { "project_id": prop_int("Project ID") },
                "required": ["project_id"]
            }),
        },
        Tool {
            name: "get_waiting_tasks".into(),
            description: "Return all tasks currently waiting in the execution queue.".into(),
            input_schema: json!({
                "type": "object",
                "properties": { "project_id": prop_int("Project ID") },
                "required": ["project_id"]
            }),
        },
        Tool {
            name: "bulk_stop_tasks".into(),
            description: "Stop multiple tasks at once by providing a list of task IDs.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "task_ids": { "type": "array", "items": { "type": "integer" }, "description": "List of task IDs to stop" }
                },
                "required": ["project_id","task_ids"]
            }),
        },
        Tool {
            name: "confirm_task".into(),
            description: "Confirm (approve) a task waiting for human approval (gated execution).".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "task_id": prop_int("Task to approve")
                },
                "required": ["project_id","task_id"]
            }),
        },
        Tool {
            name: "reject_task".into(),
            description: "Reject a task waiting for human approval — it will be stopped.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "task_id": prop_int("Task to reject")
                },
                "required": ["project_id","task_id"]
            }),
        },
    ]
}

pub async fn call(name: &str, args: &Value, client: &VelumClient) -> Result<ToolResult> {
    let pid = args.get("project_id").and_then(Value::as_i64).unwrap_or_default();
    match name {
        "list_tasks" => {
            let all = client.get(&format!("/projects/{pid}/tasks")).await?;
            let limit = args.get("limit").and_then(Value::as_u64).unwrap_or(20) as usize;
            let status_filter = args.get("status").and_then(Value::as_str);
            let tasks = all.as_array().map(|a| {
                a.iter()
                    .filter(|t| {
                        status_filter
                            .map(|s| t["status"].as_str() == Some(s))
                            .unwrap_or(true)
                    })
                    .take(limit)
                    .cloned()
                    .collect::<Vec<_>>()
            });
            Ok(ToolResult::json(&json!(tasks)))
        }
        "get_task" => {
            let tid = args["task_id"].as_i64().unwrap_or_default();
            let v = client.get(&format!("/projects/{pid}/tasks/{tid}")).await?;
            Ok(ToolResult::json(&v))
        }
        "run_task" => {
            let tid = args["template_id"].as_i64().unwrap_or_default();
            let mut body = json!({ "template_id": tid });
            if let Some(m) = args.get("message").and_then(Value::as_str) { body["message"] = json!(m); }
            if let Some(b) = args.get("git_branch").and_then(Value::as_str) { body["git_branch"] = json!(b); }
            if let Some(a) = args.get("arguments").and_then(Value::as_str) { body["arguments"] = json!(a); }
            if let Some(i) = args.get("inventory_id").and_then(Value::as_i64) { body["inventory_id"] = json!(i); }
            let v = client.post(&format!("/project/{pid}/tasks"), &body).await?;
            Ok(ToolResult::json(&v))
        }
        "stop_task" => {
            let tid = args["task_id"].as_i64().unwrap_or_default();
            client.post_empty(&format!("/projects/{pid}/tasks/{tid}/stop")).await?;
            Ok(ToolResult::text(format!("Stop signal sent to task {tid}.")))
        }
        "get_task_output" => {
            let tid = args["task_id"].as_i64().unwrap_or_default();
            let raw = args.get("raw").and_then(Value::as_bool).unwrap_or(false);
            let path = if raw {
                format!("/projects/{pid}/tasks/{tid}/raw_output")
            } else {
                format!("/projects/{pid}/tasks/{tid}/output")
            };
            let v = client.get(&path).await?;
            let lines: Vec<String> = match &v {
                Value::String(s) => s.lines().map(String::from).collect(),
                Value::Array(arr) => arr.iter()
                    .map(|r| r.get("output").and_then(Value::as_str).unwrap_or("").to_string())
                    .collect(),
                _ => vec![v.to_string()],
            };
            let n = args.get("last_n_lines").and_then(Value::as_u64).map(|n| n as usize);
            let output = match n {
                Some(n) => lines.iter().rev().take(n).rev().cloned().collect::<Vec<_>>().join("\n"),
                None => lines.join("\n"),
            };
            Ok(ToolResult::text(output))
        }
        "filter_tasks" => {
            let all = client.get(&format!("/projects/{pid}/tasks")).await?;
            let limit = args.get("limit").and_then(Value::as_u64).unwrap_or(50) as usize;
            let tpl_filter = args.get("template_id").and_then(Value::as_i64);
            let status_filter = args.get("status").and_then(Value::as_str);
            let tasks: Vec<Value> = all.as_array().map(|a| {
                a.iter()
                    .filter(|t| {
                        tpl_filter.map(|id| t["template_id"].as_i64() == Some(id)).unwrap_or(true)
                            && status_filter.map(|s| t["status"].as_str() == Some(s)).unwrap_or(true)
                    })
                    .take(limit)
                    .cloned()
                    .collect()
            }).unwrap_or_default();
            Ok(ToolResult::json(&json!(tasks)))
        }
        "get_latest_failed_task" => {
            let all = client.get(&format!("/projects/{pid}/tasks")).await?;
            let failed = all.as_array().and_then(|a| {
                a.iter()
                    .filter(|t| t["status"].as_str() == Some("error"))
                    .max_by_key(|t| t["id"].as_i64().unwrap_or(0))
                    .cloned()
            });
            match failed {
                Some(t) => Ok(ToolResult::json(&t)),
                None => Ok(ToolResult::text("No failed tasks found in this project.")),
            }
        }
        "get_waiting_tasks" => {
            let all = client.get(&format!("/projects/{pid}/tasks")).await?;
            let waiting: Vec<Value> = all.as_array().map(|a| {
                a.iter().filter(|t| t["status"].as_str() == Some("waiting")).cloned().collect()
            }).unwrap_or_default();
            Ok(ToolResult::json(&json!(waiting)))
        }
        "bulk_stop_tasks" => {
            let ids = args["task_ids"].as_array().cloned().unwrap_or_default();
            let mut results = Vec::new();
            for id_val in &ids {
                if let Some(tid) = id_val.as_i64() {
                    match client.post_empty(&format!("/projects/{pid}/tasks/{tid}/stop")).await {
                        Ok(_) => results.push(format!("✓ {tid}")),
                        Err(e) => results.push(format!("✗ {tid}: {e}")),
                    }
                }
            }
            Ok(ToolResult::text(format!("Stop results:\n{}", results.join("\n"))))
        }
        "confirm_task" => {
            let tid = args["task_id"].as_i64().unwrap_or_default();
            client.post_empty(&format!("/projects/{pid}/tasks/{tid}/confirm")).await?;
            Ok(ToolResult::text(format!("Task {tid} confirmed and will proceed.")))
        }
        "reject_task" => {
            let tid = args["task_id"].as_i64().unwrap_or_default();
            client.post_empty(&format!("/projects/{pid}/tasks/{tid}/reject")).await?;
            Ok(ToolResult::text(format!("Task {tid} rejected and stopped.")))
        }
        _ => Ok(ToolResult::error(format!("Unknown tool: {name}"))),
    }
}
